### ISSUE TITLE
StreamReader.ReadLineAsync() blocks on empty stdin in .NET 7 RC 2

### ISSUE THREAD

**Issue by k15tfu** [dotnet/runtime#77358]
Hi!

I just faced with the problem that after upgrading to .NET 7 RC 2 the following program hangs forever, but it successfully exits with a timeout in .NET 6:
```
using System;
using System.IO;
using System.Threading;

namespace readline_async_csharp
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("starting...");

            var cts = new CancellationTokenSource();
            cts.CancelAfter(3000);

            Console.WriteLine("cts will be cancelled after 3sec...");
            TextReader reader = new StreamReader(Console.OpenStandardInput(), Console.InputEncoding);
            var readTask = reader.ReadLineAsync();

            Console.WriteLine("waiting...");
            try
            {
                readTask.Wait(cts.Token);
            }
            catch (System.OperationCanceledException)
            {
                // nothing
            }

            Console.WriteLine("exiting...");
        }
    }
}
```

In .NET 6.0.10 the output is:
```
.../dotnet-sdk-6.0.402-linux-x64$ ./dotnet .../readline_async_csharp/bin/Debug/net6.0/readline_async_csharp.dll
starting...
cts will be cancelled after 3sec...
waiting...
exiting...
```

But in .NET 7.0.0-rc.2 the output is:
```
.../dotnet-sdk-7.0.100-rc.2.22477.23-linux-x64$ ./dotnet .../readline_async_csharp/bin/Debug/net7.0/readline_async_csharp.dll
starting...
cts will be cancelled after 3sec...
(still running)
```

and at this time the main thread is blocked here:
```
Main Thread @28606
	Interop.Sys.ReadStdin()
	StdInReader.ReadKey()
	StdInReader.ReadLineCore()
	StdInReader.ReadLine()
	ConsoleStream.ReadAsync()
	StreamReader.<ReadBufferAsync>d__72.MoveNext()
	AsyncMethodBuilderCore.Start<System.IO.StreamReader.<ReadBufferAsync>d__72>()
	StreamReader.ReadBufferAsync()
	StreamReader.<ReadLineAsyncInternal>d__63.MoveNext()
	AsyncMethodBuilderCore.Start<System.IO.StreamReader.<ReadLineAsyncInternal>d__63>()
	StreamReader.ReadLineAsyncInternal()
	StreamReader.ReadLineAsync()
	StreamReader.ReadLineAsync()
	Program.Main() at .../readline_async_csharp/Program.cs:line 18
```

[readline_async_csharp.tar.gz](https://github.com/dotnet/runtime/files/9847214/readline_async_csharp.tar.gz)

**Comment by MihaZupan** [comment]
Simpler repro:
```c#
Stream s = Console.OpenStandardInput();
Console.WriteLine("Before ReadAsync");
ValueTask<int> readTask = s.ReadAsync(new byte[1]);
Console.WriteLine("After ReadAsync");
await readTask;
```

The `ConsoleStream` is blocking in the `ReadAsync` call and never "goes async".
https://github.com/dotnet/runtime/blob/33cb335386d39da8e4e89e75449628be7e3b375a/src/libraries/System.Console/src/System/IO/ConsoleStream.cs#L105

This is a regression from #71971 cc: @eerhardt

**Comment by stephentoub** [comment]
> This is a regression from https://github.com/dotnet/runtime/pull/71971

Yeah, I should have noticed this when reviewing. There's zero guarantee these calls will complete promptly, even for writes, and in fact it's likely they won't. Thus they mustn't be implemented in terms of the sync APIs. This needs to be reverted and serviced asap.

**Comment by eerhardt** [comment]
This has been fixed in `main` and has been backported to `release/7.0`. It will be released in the next servicing release of `7.0.x`.
